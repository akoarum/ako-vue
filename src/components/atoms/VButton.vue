<template>
  <button :class="styledClasses" class="vButton" @click="$emit('click')">
    <slot />
  </button>
</template>

<script lang="ts">
import Vue from 'vue'

export default Vue.extend({
  props: {
    primary: { type: Boolean },
    secondary: { type: Boolean },
    caution: { type: Boolean },
    outline: { type: Boolean },
    small: { type: Boolean }
  },
  computed: {
    styledClasses(): string[] {
      const classes: string[] = []
      if (this.primary) classes.push('-primary')
      if (this.secondary) classes.push('-secondary')
      if (this.caution) classes.push('-caution')
      if (this.outline) classes.push('-outline')
      if (this.small) classes.push('-small')
      return classes
    }
  }
})
</script>

<style lang="scss" scoped>
.vButton {
  --color: #{$color-base};
  display: block;
  box-sizing: border-box;
  background: var(--color);
  line-height: 1;
  outline: none;
  color: #fff;
  transition: background 0.3s $easing;

  &:hover,
  &:focus {
    background: lighten($color-base, 15%);

    &.-outline {
      background: rgba(0, 0, 0, 0.14);
    }
  }

  @include media-min {
    padding: 9px 14px;
    min-width: 134px;
    font-size: $fontSize-pc;
  }

  @include media-max {
    padding: 9px 12px;
    width: 100%;
    font-size: $fontSize-sp;
  }
}

.-small {
  @include media-min {
    min-width: 0;
  }
}

.-outline {
  box-shadow: 0 0 0 2px currentColor inset;
  background: transparent;
  color: var(--color);
}

.-primary {
  --color: #{$color-theme};

  &:hover,
  &:focus {
    background: lighten($color-theme, 15%);

    &.-outline {
      background: lighten($color-theme, 44%);
    }
  }
}

.-secondary {
  --color: #{$color-secondary};

  &:hover,
  &:focus {
    background: lighten($color-secondary, 15%);

    &.-outline {
      background: lighten($color-secondary, 58%);
    }
  }
}

.-caution {
  --color: #{$color-caution};

  &:hover,
  &:focus {
    background: lighten($color-caution, 15%);

    &.-outline {
      background: lighten($color-caution, 52%);
    }
  }
}
</style>
