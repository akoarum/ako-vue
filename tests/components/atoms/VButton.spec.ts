import { shallowMount } from '@vue/test-utils'
import VButton from '~/components/atoms/VButton.vue'

describe('VButton', () => {
  let wrapper: any

  beforeEach(() => {
    wrapper = shallowMount(VButton, {
      propsData: {},
      slots: {
        default: 'ボタン'
      }
    })
  })

  describe('initialized', () => {
    it('mountable', () => {
      expect(wrapper.isVueInstance()).toBe(true)
    })

    it('snapshot', () => {
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  describe('computed', () => {
    it('propsを基にクラス配列を生成して返す', () => {
      expect(wrapper.vm.styledClasses).toEqual([])

      wrapper.setProps({ primary: true })
      expect(wrapper.vm.styledClasses).toEqual(['-primary'])

      wrapper.setProps({ primary: false, secondary: true })
      expect(wrapper.vm.styledClasses).toEqual(['-secondary'])

      wrapper.setProps({ secondary: false, caution: true })
      expect(wrapper.vm.styledClasses).toEqual(['-caution'])

      wrapper.setProps({ outline: true })
      expect(wrapper.vm.styledClasses).toEqual(['-caution', '-outline'])

      wrapper.setProps({ small: true })
      expect(wrapper.vm.styledClasses).toEqual(['-caution', '-outline', '-small'])
    })
  })
})
