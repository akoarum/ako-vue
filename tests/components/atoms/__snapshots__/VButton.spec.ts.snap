// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VButton initialized snapshot 1`] = `
<button
  class="vButton"
>
  ボタン
</button>
`;
